### PR TITLE
#1952 allow hiding voters in polls / bug-fixes for #1953, #1959

### LINKED ISSUE #1952: Allow hiding voters in polls
### Problem

Some questions are better answered anonymously.

### Solution

Add a checkbox called `Hide who voted` to the poll creation box in the Chat to mark a poll as being "anonymous"*.

If a poll is marked as anonymous, do not show who voted on it in the UI.

_*Note: these are not completely anonymous votes. It is still possible for someone skilled to "hack" the app and figure out who voted on what. To create a truly cryptographically anonymous voting system is beyond the scope of this issue._

### LINKED ISSUE #1953: Error in sideEffect related to sending a poll in a DM
### Problem

Screenshot 2024-04-19 at 9 04 09 AM

@dotmacro sent me a DM with the poll below, and I *think* that's what triggered triggered the error above.

Screenshot 2024-04-19 at 9 11 02 AM

Here's the corresponding error log:

```js
{
  "timestamp": "2024-04-19T16:03:10.260Z",
  "type": "error",
  "msg": [
    "[chelonia] ERROR 'TypeError' in sideEffect for <op_ae|z9brRu3VMmmxg1de3jjiwcjUaQDALmaBUJ2QXU8GYS4vEcKqebGx of z9brRu3VUgFkg1eAfirGJ454n49S9sDgBDRx3EDAzPHXzUMSjaXj>: Cannot read properties of undefined (reading 'split')",
    {
      "name": "TypeError",
      "message": "Cannot read properties of undefined (reading 'split')",
      "stack": "TypeError: Cannot read properties of undefined (reading 'split')\n    at swapUserIDForUsername (eval at chelonia/private/loadManifest (https://xxx.xxx.xxx/assets/js/chunk-M5ENVJQB-cached.js:1:19965), <anonymous>:7074:17)\n    at messageReceivePostEffect (eval at chelonia/private/loadManifest (https://xxx.xxx.xxx/assets/js/chunk-M5ENVJQB-cached.js:1:19965), <anonymous>:7198:13)\n    at Object.sideEffect (eval at chelonia/private/loadManifest (https://xxx.xxx.xxx/assets/js/chunk-M5ENVJQB-cached.js:1:19965), <anonymous>:7475:11)\n    at z9brRu3VQeSP1LaQiXp7p55UkU4rfrSBzkYeEyDQ4daQYyoQcSsg/gi.contracts/chatroom/addMessage/sideEffect (https://xxx.xxx.xxx/assets/js/chunk-M5ENVJQB-cached.js:22:22486)\n    at u (https://xxx.xxx.xxx/assets/js/chunk-SSANYCJJ-cached.js:1:222)\n    at l (https://xxx.xxx.xxx/assets/js/chunk-M5ENVJQB-cached.js:19:27723)\n    at Object.processSideEffects (https://xxx.xxx.xxx/assets/js/chunk-M5ENVJQB-cached.js:19:27802)\n    at chelonia/private/in/handleEvent (https://xxx.xxx.xxx/assets/js/chunk-M5ENVJQB-cached.js:19:25051)\n    at async https://xxx.xxx.xxx/assets/js/chunk-M5ENVJQB-cached.js:19:13995\n    at async okTurtles.eventQueue/queueEvent (https://xxx.xxx.xxx/assets/js/chunk-M5ENVJQB-cached.js:1:2532)"
    },
    {
      "message": "{\"_signedData\":[\"[\\\"z2Drjgb9hxXGP7dt5AtgrviEuorDiJWvvyNStfTfSdUn29vPCje\\\",\\\"8Llpy+slYCKqOgv4rknr3/KTwlLhlIK91Oqlu26VUQSZAzksM1mkIB9q3wbwAQb+85RC6JTP06pZiTSqakN7GDDdwrnbC2tyDypVetYPE9UxMty15xMAKX1zrU2Tbx/GW/+P7rHqlb3nYwEawGXnYf+fNhXel79GWC9et4PQyuLc2hgBOfqk3/t2xSqFCj+OQ/SOSaZxH22kTon1GWlmurMJZbdwd2Ul5VM1nJDYGRe/V6Kh0Ses/FLeb++ZKLcbFECwjvB5LE3n2nH9LbpIvpIJwE/zE8S2VfW0U9iAeqpQv31E2DPF5BloUztgk6/vohmQUnq+OxwWVeVLYjoauOhIDQVreVgm7wnGneCvZVbdNVHkFcBJDFRF8PSgdT/6eucfXhV9W6iKWa9dqtSB/lIe9vJ7ClDnbJWMPp+OELJNpCb30fsQqy01aQqP2YhVDUqhVclspNzoPMbUBS7uYHgiR7xXqaaaZINcIpL2csSNMY5j9fZw72c6+Q8u2Iag5kjGM+p9wRzmDV59GVlfPy3hbgB/90PpMZaZXVzsxaj8rn1eE/rskZgGVwBsQmWNOJrQFSCABt9HALftNXy0Y9Pvouo40EUAuI90Vy37C1CIAJF1VDVSyn/MvaRlNw5zNECW7KP7j9Va8pVD6jKrtFJzyiKZNgYshCRmdaJTTt5SbCjcGb5PH8SJDCnGI3MA1diIz+S/v2xjifEhIvnoQMsTZDv8M+QlGe1CQf4SYlrglyE//aizlvzULs5I3GaE6ZU53KzHmdrFN6jj1Dt/B5znNTh/C1DhU0TJ5SB+mzJciFt0FgV2YAxYNJb7zJwgDWfXxl9cVUmoOswgRtT2suM0qUsvHcns0LBd15VVVxlanChnaVsCPM3AjX+ergpLNHJmfSLU9CsmzjxLVq0mwku9zNdezs5rXpKeDfxzMyV+CIeuf29gE8Air0dOfkbgwDV5mtZCfZ+7t8tmvhPXMb4s4oGHYkpDJ0LlqNkTMfcqjFAqp568f1Pj8w3u64Xi+p+7T8COCDfVP9YCQD0Mlrf3X5TohvTbMJwN7DRN6+gTkTDzZezEFQ==\\\"]\",\"z2DrjgbJf8CxGUDyxr1RPNM5dHHihbvZ5HAPHFmU5wFiAZuSJ3n\",\"VfveWKcoDrpPA2/yNoH9vzD5ABk5MKur6iIK2qfIAn0PCPY7OBsj+3/g+MGlaHYplvfdFVnncfq52L+ZlOLKBw==\"],\"head\":\"{\\\"version\\\":\\\"1.0.0\\\",\\\"previousHEAD\\\":\\\"z9brRu3VRgcQe9iGi5db6VeXm7cJCAsS42DxNd1stZYg79PFSa4H\\\",\\\"height\\\":8,\\\"contractID\\\":\\\"z9brRu3VUgFkg1eAfirGJ454n49S9sDgBDRx3EDAzPHXzUMSjaXj\\\",\\\"op\\\":\\\"ae\\\",\\\"manifest\\\":\\\"z9brRu3VQeSP1LaQiXp7p55UkU4rfrSBzkYeEyDQ4daQYyoQcSsg\\\"}\"}"
    }
  ]
}
```

NOTE: Interestingly enough, I only got this error while I was logged in on Desktop to the app. I did not see it later in the mobile PWA version of the app.

### Solution

Find and fix the problem!

### LINKED ISSUE #1959: Vote notification in a DM chatroom has a blank text
### Problem

When a poll is created as part of a DM chatroom and the members vote on the poll, the `MessageNotification.vue` component produced afterwards don't have text saying, `voted on {option 1}` or `changed votes to {option 2}` etc.

### Solution
Find the root cause and fix it.

### PR DESCRIPTION
**Pull request by SebinSong**
closes #1952 
closes #1953 
closes #1959 

Added UI updates are as below:

**1) The checkbox option with a disclaimer**

**2) '(Anonymous)' will be added to the message UI**

**3) Users cannot see who voted on what**

**4) Vote notification messages don't say which options were voted on but only says 'Voted' or 'Changed vote'**

---
Hope they all look good.

One thing I wasn't sure about was **4.** above. Pls let me know if the notification messages shouldn't be generated at all, rather than it hides the options selected. 

cc. @taoeffect

### CODE REVIEW

File: frontend/views/containers/proposals/ProposalItem.vue [hunk 1]
```diff
@@ -309,7 +309,7 @@ export default ({
         if (
```
Comment (SebinSong) [inline frontend/views/containers/proposals/ProposalItem.vue]: the changes in this file is not related to the issue itself but was a fix for a test-failure (in `group-proposals.spec.js`) in `master` branch FYI.

**Review by taoeffect — CHANGES_REQUESTED** [review]
Nice!

> Pls let me know if the notification messages shouldn't be generated at all, rather than it hides the options selected.

Indeed, before I do an in-depth code review, could you please make the following changes:

- Replace "(Anonymous)" with "(Votes Hidden)"
- Remove all messages from the chat when someone votes (both "u1 Voted" and "u2 Changed vote" type messages)

File: frontend/model/contracts/chatroom.js [hunk 2]
```diff
@@ -107,7 +107,7 @@ function messageReceivePostEffect ({
 
   shouldNotifyMessage && makeNotification({
     title,
-    body: swapUserIDForUsername(text),
+    body: messageType === MESSAGE_TYPES.TEXT ? swapUserIDForUsername(text) : L('New message'),
```
Comment (SebinSong) [inline frontend/model/contracts/chatroom.js]: @taoeffect 
This is the fix for the issue #1953 .
The cause of the issue was that the `text` here is only a valid value when the message is of type `MESSAGE_TYPES.TEXT`. for the rest of the types (e.g. `MESSAGE_TYPES.POLL`) it is `undefined` and in these case, the browser notification for the DM chat will say `New message` as the default. But let me know if you have a better idea though.

[MESSAGE_TYPES.TEXT]


[MESSAGE_TYPES.POLL]

**Comment by SebinSong** [comment]
As mentioned in the PR description above too, the PR fixes the issue #1959 .

The DM now properly shows the vote notification messages like below

**Review by taoeffect — CHANGES_REQUESTED** [review]
Hopefully I'll be able to start my review on this one tomorrow morning, but in the meantime I noticed that this PR pins a new version of the contracts.

Could you revert everything related to that? (delete `contracts/0.2.4` and all mention of `0.2.4` from `package.json`). The way we release now is we have a `releases` branch, and only there do we bump the contracts version.

File: frontend/views/containers/proposals/ProposalItem.vue [hunk 3]
```diff
@@ -326,7 +326,7 @@ export default ({
         !this.currentGroupState._vm?.authorizedKeys?.[inviteKeyId] ||
         // If _notAfterHeight is *not* undefined, it means that the key has been
         // revoked. Hence, it cannot be used
-        this.currentGroupState._vm.authorizedKeys[inviteKeyId]._notAfterHeight !== undefined ||
```
Comment (taoeffect) [inline frontend/views/containers/proposals/ProposalItem.vue]: Are you certain about this change? Is this something @corrideat missed?

Comment (SebinSong) [reply frontend/views/containers/proposals/ProposalItem.vue]: I think this was a fix for a failing cypress test at that time which is unrelated to this work. If that has been fixed in the `master`, I will put it back to whatever it is there.

Comment (SebinSong) [reply frontend/views/containers/proposals/ProposalItem.vue]: changed it according to `master`

**Review by taoeffect — CHANGES_REQUESTED** [review]
When I created a hidden poll in a DM (**not** multiple choice), this is how it appears in the DM:

Screenshot 2024-05-07 at 4 01 26 PM

- NaN% shouldn't be there
- It doesn't let anyone vote either
- And for some reason says "Poll Closed"

**Comment by SebinSong** [comment]
@taoeffect That's wierd..
This is how I look for Polls created in DM.

image

Did you test this PR in the correct feature branch of mine which is `sebin/task/#1952-allow-hiding-voters-in-polls`?

**Comment by taoeffect** [comment]
@SebinSong Yes, it's using the latest commit 121943bfc9d845437ba1367c6b4d3c1cd4b4cd47. It seems to not always happen...

For example, I just did this:

1. [Tab1] create group
2. [Tab2] u2 join, create chatroom
3. [Tab2] u2 create DM, create and send hidden poll to u1 (it works!)
4. [Tab2] u2 go to chatroom that was created in step (2) and post another hidden poll (it doesn't work!)

Screenshot 2024-05-07 at 6 16 17 PM

**Comment by SebinSong** [comment]
@taoeffect 
That was right. It happens when trying to create another Poll subsequently after creating initial one.
The reason was that when the `this.form` object is reset after successfully creating a Poll message, it misses `duration` field in there and that leads to `expires_date_ms` value of any subsequent Poll data set to `NaN`, which leads to the immediate expiration. made a fix [here](https://github.com/okTurtles/group-income/pull/1964/commits/002655e4f482df6dbc567b122537c573006422f2) accordingly.

Your screenshot actually revealed another issue which is **the graph displaying `NaN%` when a Poll is expired with no one ever voting on it.** The percentage calculation needed to take this case into account so I made an update for it too.
Now it correctly says 0% like below.

**Review by taoeffect — APPROVED** [review]
Nicely done 👍

**Pull request merged**